apps/emqx_rule_engine/src/emqx_rule_engine_api.erl

```diff
@@ -250,7 +250,8 @@ do_create_rule2(ParsedParams) ->
             return({error, 400, ?ERR_BADARGS(Reason)})
     end.
 
-update_rule(#{id := Id}, Params) ->
+update_rule(#{id := Id0}, Params) ->
+    Id = urldecode(Id0),
```

Comment:
I have a question, will this work correctly with user's old data?

Reply:
#### yes it is.
All literal value will be same as it after urldecoded.
i.e.
```
(emqx@127.0.0.1)2> emqx_http_lib:uri_decode(<<"rule:1234">>).
<<"rule:1234">>
(emqx@127.0.0.1)3> emqx_http_lib:uri_decode(<<"words with space">>).
<<"words with space">>
(emqx@127.0.0.1)4> emqx_http_lib:uri_decode(<<"words with space but some%20of%20them%20uri%20encoded">>).
<<"words with space but some of them uri encoded">>
```

#### So it only effects the rule id contains percent `%` in case id contains exactly a value that can be percent-encoded. i.e. `%20`, `%25`

In this case, if a rule was created with id `%25` (literal value) in post http body on old versions. Obviously it **can't** be shown by http request `GET /api/v4/rule/%2525`. or `GET /api/v4/rule/%25`.
But now `%2525` in path(percent-encoded id) will be urldecoded to `%25`(literal value, the real id)